src/mito2/src/sst/index/fulltext_index/creator.rs

```diff
@@ -0,0 +1,293 @@
+                    intm_path,
+                    compress,
+                },
+            );
+        }
+
+        Ok(Self {
+            creators,
+            aborted: false,
+            stats: Statistics::default(),
+        })
+    }
+
+    /// Updates the index with the given batch.
+    pub async fn update(&mut self, batch: &Batch) -> Result<()> {
+        ensure!(!self.aborted, OperateAbortedIndexSnafu);
+
+        if let Err(update_err) = self.do_update(batch).await {
+            if let Err(err) = self.do_abort().await {
+                if cfg!(any(test, feature = "test")) {
+                    panic!("Failed to abort index creator, err: {err}");
+                } else {
+                    warn!(err; "Failed to abort index creator");
+                }
+            }
+            return Err(update_err);
+        }
+
+        Ok(())
+    }
+
+    /// Finalizes the index creation.
+    pub async fn finish(
+        &mut self,
+        puffin_writer: &mut SstPuffinWriter,
+    ) -> Result<(RowCount, ByteCount)> {
+        ensure!(!self.aborted, OperateAbortedIndexSnafu);
+
+        match self.do_finish(puffin_writer).await {
+            Ok(()) => Ok((self.stats.row_count(), self.stats.byte_count())),
+            Err(finish_err) => {
+                if let Err(err) = self.do_abort().await {
+                    if cfg!(any(test, feature = "test")) {
+                        panic!("Failed to abort index creator, err: {err}");
+                    } else {
+                        warn!(err; "Failed to abort index creator");
+                    }
+                }
+                Err(finish_err)
+            }
+        }
+    }
+
+    /// Aborts the index creation.
+    pub async fn abort(&mut self) -> Result<()> {
+        if self.aborted {
+            return Ok(());
+        }
+
+        self.do_abort().await
+    }
+
+    /// Returns the memory usage of the index creator.
+    pub fn memory_usage(&self) -> usize {
+        self.creators.values().map(|c| c.inner.memory_usage()).sum()
+    }
+
+    /// Returns IDs of columns that the creator is responsible for.
+    pub fn column_ids(&self) -> impl Iterator<Item = ColumnId> + '_ {
+        self.creators.keys().copied()
+    }
+
+    pub(crate) fn is_empty(&self) -> bool {
+        self.creators.is_empty()
+    }
+}
+
+impl SstIndexCreator {
+    async fn do_update(&mut self, batch: &Batch) -> Result<()> {
+        let mut guard = self.stats.record_update();
+        guard.inc_row_count(batch.num_rows());
+
+        for creator in self.creators.values_mut() {
+            creator.update(batch).await?;
+        }
+
+        Ok(())
+    }
+
+    async fn do_finish(&mut self, puffin_writer: &mut SstPuffinWriter) -> Result<()> {
+        let mut guard = self.stats.record_finish();
+
+        let mut written_bytes = 0;
+        for creator in self.creators.values_mut() {
+            written_bytes += creator.finish(puffin_writer).await?;
+        }
```

Comment:
Prefer the iterator with reduce and sum, but it's ok.

Reply:
Clause with async and &mut ref is annoying. The following code is closed but still doesn't work due to ref of `puffin_writer`.

```rust
let written_bytes = stream::iter(self.creators.values_mut())
    .then(|creator| async move { creator.finish(puffin_writer).await })
    .try_fold(0, |acc, x| async move { Ok(acc + x) })
    .await?;
```